### PR TITLE
Active users cruncher with a simple test set

### PR DESCRIPTION
**Pull request by grixa**
CC @dkorolev @mzhurovich

### CODE REVIEW

File: crunchers/cruncher_active_users.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+  uint64_t Count() const { return users_list_.size(); }
+  uint64_t Size() const { return stream_size_; }
+
+  void operator()(const typename NAMESPACE::EventLogEntry& e) {
+    current_us_ = e.server_us;
+    e.event.Call(*this);
+  }
+
+  void operator()(const typename NAMESPACE::iOSGenericEvent& e) { OnIOSEvent(e); }
```
Comment (dkorolev) [inline crunchers/cruncher_active_users.h]: `template<>`, `enable_if<>`, `is_base_of<>` instead of copy-pasting?

Comment (dkorolev) [inline crunchers/cruncher_active_users.h]: Or even call `Wrapper`, not `*this`, which could just have a `template<T>`, eliminating the need in `enable_if`?

Comment (dkorolev) [inline crunchers/cruncher_active_users.h]: And the template is not even needed: `Exists<iOSBaseEvent>`/`Value<iOSBaseEvent>` should do just fine!

Comment (dkorolev) [inline crunchers/cruncher_active_users.h]: (instead of `.Call(*this)` I mean.)

File: crunchers/test.cc [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+  std::unique_ptr<CTFOActiveUsersCruncher> activeusers_cruncher;
+
+  const auto SubscribeAndWait = [&]() {
+    const auto scope = local_stream.Subscribe(*activeusers_cruncher);
+    while (activeusers_cruncher->Size() < 80) {
+      std::this_thread::yield();
+    }
+  };
+
+  // In the last 12 seconds there were 12 users, which appeaed exactly one per second.
```
Comment (mzhurovich) [inline crunchers/test.cc]: Typo: appea**r**ed

File: crunchers/test.cc [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  }
+
+  // In the previous 12 minutes there were no 'new' users, only the same ones.
+  for (uint32_t i = 1; i <= 13; ++i) {
+    activeusers_cruncher = std::make_unique<CTFOActiveUsersCruncher>(std::chrono::minutes(i));
+    SubscribeAndWait();
+    EXPECT_EQ(12u, activeusers_cruncher->Count());
+  }
+
+  // A minute berfore there was one more 'unique' active user.
```
Comment (mzhurovich) [inline crunchers/test.cc]: Typo: berfore->before

File: crunchers/test.cc [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+  activeusers_cruncher = std::make_unique<CTFOActiveUsersCruncher>(std::chrono::minutes(14));
+  SubscribeAndWait();
+  EXPECT_EQ(13u, activeusers_cruncher->Count());
+
+  // And a minute berfore that - another one.
```
Comment (mzhurovich) [inline crunchers/test.cc]: same here

File: crunchers/test.cc [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    SubscribeAndWait();
+    EXPECT_EQ(12u, activeusers_cruncher->Count());
+  }
+  
+  // A minute berfore there was one more 'unique' active user.
+  activeusers_cruncher = std::make_unique<CTFOActiveUsersCruncher>(std::chrono::minutes(14));
+  SubscribeAndWait();
+  EXPECT_EQ(13u, activeusers_cruncher->Count());
+  
+  // And a minute berfore that - another one.
```
Comment (mzhurovich) [inline crunchers/test.cc]: All the same typos in this test :)

File: crunchers/cruncher_active_users.h [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    if (Exists<EventLogEntry>(e)) {
+      OnEventLogEntry(Value<EventLogEntry>(e));
+    }
+    while (!users_list_.empty() && users_list_.back().us + interval_ <= current_us_) {
+      users_map_.erase(users_list_.back().device_id);
+      users_list_.pop_back();
+    }
+  }
+
+  uint64_t Count() const { return users_list_.size(); }
```
Comment (mzhurovich) [inline crunchers/cruncher_active_users.h:57]: `OnEvent` method above and `Count` are supposed to be called from different threads, right? What about synchronization primitives?

File: crunchers/cruncher_active_users.h [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+      users_list_.push_front(ActiveUser{current_us_, e.device_id});
+      users_map_[e.device_id] = users_list_.begin();
+    }
+  }
+
+  user_list_t users_list_;
+  user_map_t users_map_;
+  std::chrono::microseconds current_us_;
+  std::chrono::microseconds interval_;
+  uint64_t stream_size_;
```
Comment (mzhurovich) [inline crunchers/cruncher_active_users.h]: This one looks more like `events_seen_` for me.

**Pull request merged**